File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#
from typing import List
# @lc code=start
class Solution:

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []

        ans = set()

        for e in nums:
            res = self.twoSum(nums, -e)
            if res:
                res.append(e)
                ans.add(tuple(sorted(res)))

        return ans

    def twoSum(self, nums, target):
        seen = {}
        for e in nums:
            remaining = target - e
            if remaining in seen:
                return [seen[remaining], e]
            seen[e] = e
        return []

    def threeSumOfficial(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        ans = list()

        # 枚举 a
        for first in range(n):
            # 需要和上一次枚举的数不相同
            if first > 0 and nums[first] == nums[first - 1]:
                continue
            # c 对应的指针初始指向数组的最右端
            third = n - 1
            target = -nums[first]
            # 枚举 b
            for second in range(first + 1, n):
                # 需要和上一次枚举的数不相同
                if second > first + 1 and nums[second] == nums[second - 1]:
                    continue
                # 需要保证 b 的指针在 c 的指针的左侧
                while second < third and nums[second] + nums[third] > target:
                    third -= 1
                # 如果指针重合，随着 b 后续的增加
                # 就不会有满足 a+b+c=0 并且 b<c 的 c 了，可以退出循环
                if second == third:
                    break
                if nums[second] + nums[third] == target:
                    ans.append([nums[first], nums[second], nums[third]])

        return ans

# @lc code=end
s = Solution()
s.threeSum([0, 0 ,0])

File: pointers/15.3-sum.go
package leetcode

import "sort"

// 排序+双指针
func threeSum(nums []int) [][]int {
	n := len(nums)
	sort.Ints(nums)
	ans := make([][]int, 0)

	// 固定一个 a
	for first := 0; first < n; first++ {
		if first > 0 && nums[first] == nums[first-1] { // 规避重复
			continue
		}
		third := n - 1 // c 在最右端
		target := -1 * nums[first]
		// 枚举 b、c，分别位于寻找范围的两侧
		for second := first + 1; second < n; second++ {
			if second > first+1 && nums[second] == nums[second-1] { // 规避重复
				continue
			}

			// b 在 c 的左边，找目标
			for second < third && nums[second]+nums[third] > target {
				third--
			}
			if second == third {
				break // 没找到
			}
			if nums[second]+nums[third] == target {
				ans = append(ans, []int{nums[first], nums[second], nums[third]})
			}
		}
	}
	return ans
}
